### PR TITLE
Partial cuIO GPU decompression refactor

### PR DESCRIPTION
**Pull request by vuule**
Required to expand future nvcomp integration.

- [x] Moving nvcomp integration in ORC and Parquet readers to common code. Enables nvcomp use for multiple compression type without code duplication.
- [x] `gpu_inflate_input_s` refactor to facilitate unified host/device decompressor interface. Enables further changes to unify CPU and GPU decompression API, which in turn enables ZSTD use in ORC.

### CODE REVIEW

File: cpp/src/io/comp/gpuinflate.h
Comment (vuule) [inline cpp/src/io/comp/gpuinflate.h:30]: Been thinking about this, and I'm still not happy with the name of this struct:

1. Turns out it's also used for compression. 
2. It's just a pair of device buffer views, there's nothing specific to (de)compression.

Any suggestions?

File: cpp/src/io/comp/unsnap.cu
Comment (bdice) [inline cpp/src/io/comp/unsnap.cu:636]: (This feels like a case where a `reinterpret_cast` has done something very, very unsafe?)

What does `reinterpret_cast<const uint32_t*>(&inputs[strm_id])[t]` mean in this expression? It doesn't refer to the `src` or `dst` members. Is this reinterpreting a `device_span`? The `struct device_decompress_input` itself? It's getting late and my brain is melting so I'm going to wrap up my review here for now.

**Review by bdice — CHANGES_REQUESTED** [review]
Stopping halfway through this for now. I'll need to make another review pass to finish up next week.

File: cpp/src/io/comp/unsnap.cu
Comment (vuule) [reply cpp/src/io/comp/unsnap.cu:636]: This is a cooperative copy AFAICT. We removed most of these, looks like this one survived. I'll replace with a simple assignment.

File: cpp/src/io/comp/gpuinflate.h
Comment (vuule) [reply cpp/src/io/comp/gpuinflate.h:30]: That makes sense. I was thinking of merging dst and status as output, but having all of them separate is even better 👍

File: cpp/src/io/parquet/page_enc.cu
Comment (bdice) [inline cpp/src/io/parquet/page_enc.cu:84]: Should these be device spans?

**Review by bdice — COMMENTED** [review]
Only one other comment over the rest of the PR for now.

**Review by robertmaynard — APPROVED** [review]
CMake changes LGTM

File: cpp/src/io/comp/gpuinflate.h
Comment (vuule) [reply cpp/src/io/comp/gpuinflate.h:56]: the naming of older cuIO kernels follows a pattern that's not used elsewhere, reorg/renames would be good. There's also a gpu namespace in most components that we'd like to remove.
FWIW, we are aiming to eventually remove most of the code in `/comp` and use nvcomp instead.

File: cpp/src/io/comp/snap.cu
Comment (vuule) [reply cpp/src/io/comp/snap.cu:272]: yeah, I don't think poking at this is useful. Also, this code is already semi-retired and should be removed in a few releases.

**Review by elstehle — COMMENTED** [review]
Great initiative! I've focused on the parts you actually touched. Overall looks good; just two high-level questions/suggestions.

File: cpp/src/io/comp/nvcomp_adapter.cu
Comment (elstehle) [inline cpp/src/io/comp/nvcomp_adapter.cu:77]: Could this be replaced with a `thrust::transform` using a `zip_iterator` that zips together `actual_uncompressed_sizes` and `nvcomp_stats` as input? Would it improve things?

**Review by bdice — COMMENTED** [review]
The improvements from reviews made this PR look a lot better! I have some comments attached and will approve after some of the comments are addressed.

An apology in advance -- I attempted to find/fix some places where docstrings seemed to be marking `inputs` as `@param[out]`. While this may in fact be correct in some of the functions, I'm not sure that _all_ the docstrings are correct -- and I'm also not sure if I correctly fixed all of the errors. Similarly, there are a few `device_span<device_span<T const>>` that could probably be `device_span<device_span<T const> const>` (with an extra `const` on the outer span). I couldn't quite track all the functions' input/output patterns from what I saw in the diffs, and I decided it was not worth digging deeper to verify each one at the cost of delaying the review. I would suggest a very quick pass to double-check your expectations for `const`ness of inputs/outputs/statuses.

File: cpp/src/io/parquet/reader_impl.cu
Comment (vuule) [reply cpp/src/io/parquet/reader_impl.cu:1188]: `hostdevice_vector` converts to `device_span` and `host_span`. Unless I'm missing something, we're constructing spans manually to create a view to a subspan of the vector.
We could add a H2D and D2H APIs that only copy a subspan.

Summoning the Parquet reader expert @nvdbaranec for opinion on the use of `hostdevice_vector` here.

**Review by elstehle — COMMENTED** [review]
Almost through. Just a few minor comments and some nitpicks.

File: cpp/src/io/comp/debrotli.cu
Comment (vuule) [reply cpp/src/io/comp/debrotli.cu:2020]: we use `t` very consistently across cuIO kernels (for better of for worse). I've changed `z` to `block_id`, that one was odd to me as well.

File: cpp/src/io/orc/orc_gpu.h
Comment (vuule) [reply cpp/src/io/orc/orc_gpu.h:59]: I don't think they are, Would prefer to keep as-is.
BTW, this is the first time I've seen a class with "in" and "out" data members :D

File: cpp/src/io/orc/orc_gpu.h
Comment (vuule) [reply cpp/src/io/orc/orc_gpu.h:65]: This is still needed, since this is the size of the array the dec_in_ctl points to. Since it's just a pointer, we have to pass the size separately (changing it from a pointer to span would open another can of worms).
Probably not apparent, but this comment indicates that `num_compressed_blocks` changes its meaning at some point, as it is the number of entries in `dec_in_ctl` going in, but it's the number of compressed blocks coming out :D
Cleaning this up deserves a dedicated effort IMO.

File: cpp/src/io/orc/orc_gpu.h
Comment (vuule) [reply cpp/src/io/orc/orc_gpu.h:62]: It's not really consistent to use a span here, since the other members are pointers. For example, `dec_in_ctl` should be a span of span, but the code requires it to be a pointer :vomiting_face: 
I changed it into a span anyway, every bit helps with this mess :)

File: cpp/src/io/orc/orc_gpu.h
Comment (vuule) [reply cpp/src/io/orc/orc_gpu.h:364]: Can't use const here, this function actually initializes the spans, not just the contained data.

File: cpp/src/io/orc/orc_gpu.h
Comment (vuule) [reply cpp/src/io/orc/orc_gpu.h:351]: yeah, it's initialized here.

File: cpp/src/io/parquet/parquet_gpu.hpp
Comment (vuule) [reply cpp/src/io/parquet/parquet_gpu.hpp:588]: this actually is an output parameter, `gpuEncodePages` initializes the elements of `comp_in` and `comp_out`. Not the data, the actual spans.

File: cpp/src/io/utilities/hostdevice_vector.hpp
Comment (vuule) [reply cpp/src/io/utilities/hostdevice_vector.hpp:54]: This is a fun one. Once I followed your suggestion to use hostdevice_vector in the comp tests, a few errors popped up. Turns out additional warning are treated as errors in tests (or additional warnings are enabled, not sure). 
One of the warnings was that `max_elements` and `num_elements` are declared and initialized in different order, and that's theoretically a no-go in C++.
I also had to add a cast to fix the other warning.

File: cpp/src/io/parquet/page_enc.cu
Comment (vuule) [reply cpp/src/io/parquet/page_enc.cu:1097]: I didn't realize it's a global function. Removed :+1:

File: cpp/src/io/parquet/reader_impl.cu
Comment (vuule) [reply cpp/src/io/parquet/reader_impl.cu]: Made the changes to `hostdevice_vector`. The other places where the API can be used should be addressed in a follow-up PR.

File: cpp/src/io/orc/orc_gpu.h
Comment (bdice) [reply cpp/src/io/orc/orc_gpu.h]: Oh, I didn't even see the `*` indicating that `dec_in_ctl` was a pointer. Otherwise I wouldn't have asked this. 🙃

File: cpp/src/io/utilities/hostdevice_vector.hpp
Comment (bdice) [inline cpp/src/io/utilities/hostdevice_vector.hpp:127]: I'm not sure how I feel about the partial copy methods here. At first, I was thinking we would solve this problem by using a different container (e.g. `std::vector` on the host) and use [vector factories](https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/include/cudf/detail/utilities/vector_factories.hpp) with the spans of items to be copied (to an `rmm::uvector` on the device). I would prefer that approach since it seems the utility of subspan copying on the `hostdevice_vector` class is limited. I realize I didn't communicate very clearly, but this is what I had in mind when I [said](https://github.com/rapidsai/cudf/pull/10699#discussion_r858162650) "If this only copies a subset of the elements, it seems unintuitive to use a `hostdevice_vector` [...] Maybe a correctly-sized container that automatically converts to `device_span` would be better."

`hostdevice_vector` seems most appropriate when we know we want pinned memory that can copy efficiently back and forth many times. If I recall correctly, the host/device copies in this case are not happening many times in a row, so maybe having the pinned host/device container isn't really needed. Vector factories that copy a subspan of memory might be equally good for this atypical use case of copying subspans.

Also, adding these methods opens up `hostdevice_vector` to a new class of out-of-range errors where `start_pos + count > size()`. I would not want to add bounds checking to a container primitive, so documentation listing the requirements would be important if we choose to go this path.

What do you think about reverting this and using vector factories to copy host/device data from the appropriate spans? I think it keeps the scope of this PR tighter and avoids introducing single-use behavior.

Comment (bdice) [reply cpp/src/io/utilities/hostdevice_vector.hpp:127]: Aside from this issue, I'm ready to approve the PR. Happy to discuss pros/cons with you before implementing changes.

**Review by bdice — APPROVED** [review]
Discussed offline a bit with @vuule. The changes in 2bd4f4e8c696ee9a69f26a4df54b486b1990be17 are nice. 👍

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**
